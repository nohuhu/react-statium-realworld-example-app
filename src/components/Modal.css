.modal-dialog {
  position: relative;
  width: auto;
  margin: .5rem;
  pointer-events: none;
}

.modal-content {
  position: relative;
  display: flex;
  flex-direction: column;
  width: 100%;
  pointer-events: auto;
  background-color: #fff;
  background-clip: padding-box;
  border: 1px solid rgba(0,0,0,.2);
  border-radius: .3rem;
  outline: 0;
}

.modal-header {
  display: flex;
  align-items: flex-start;
  justify-content: space-between;
  padding: 1rem;
  border-bottom: 1px solid #e9ecef;
  border-top-left-radius: .3rem;
  border-top-right-radius: .3rem;
}

.modal-header .close {
  padding: 1rem;
  margin: -1rem -1rem -1rem auto;
}

.modal-title {
  margin-bottom: 0;
  line-height: 1.5;
}

.modal-footer > :not(:last-child) {
  margin-right: .25rem;
}

.modal-backdrop.show {
  opacity: .5 !important;
}

.fade {
  opacity: 0;
  transition: opacity .15s linear;
}

.fade.show {
  opacity: 1;
}

.modal.show .modal-dialog {
  transform: translate(0,0);
}

.modal.fade .modal-dialog {
  transition: transform .3s ease-out;
  transform: translate(0,-25%);
}

@media (min-width: 576px) {
  .modal-dialog-centered {
    min-height: calc(100% - (1.75rem * 2));
  }

  .modal-dialog {
    max-width: 500px;
    margin: 5rem auto 1.75rem auto;
  }
}