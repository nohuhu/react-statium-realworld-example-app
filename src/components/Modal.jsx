import React, { useEffect, useRef } from 'react';

// main.css from demo.productionready.io is based on very old
// 4.0.0-alpha2 version of Bootstrap. This stylesheet adds
// missing styles to make modals work as expected(ish).
import './Modal.css';

const showModal = modalEl => {
  document.body.classList.add('modal-open');

  const backdropEl = document.createElement('div');
  backdropEl.className = 'modal-backdrop fade show';
  document.body.appendChild(backdropEl);

  modalEl.classList.add('show');
  modalEl.style.display = 'block';
};

const hideModal = modalEl => {
  document.body.classList.remove('modal-open');

  const backdropEl = document.body.querySelector('.modal-backdrop');

  if (backdropEl) {
    document.body.removeChild(backdropEl);
  }

  modalEl.classList.remove('show');
  modalEl.style.display = 'none';
};

const Modal = props => {
  const modalRef = useRef(null);

  const {
    show,
    title,
    closable = true,
    content,
    callback = () => {},
    buttons = [],
  } = props;

  useEffect(() => {
    if (show) {
      showModal(modalRef.current);
    }
    else {
      hideModal(modalRef.current);
    }
  }, [show]);

  return (
    <div ref={modalRef} className="modal fade" tabIndex="-1" role="dialog">
      <div className="modal-dialog" role="document">
        <div className="modal-content">
          <div className="modal-header">
            <h5 className="modal-title">{title}</h5>
            {closable && (
              <button type="button"
                className="close"
                aria-label="Close"
                onClick={e => {
                  e.preventDefault();
                  // We don't pass back the button here simply because
                  // it is easier to do in the demo. Most actions calling
                  // modal() will not care for this result anyway.
                  callback(null);
                }}
              >
                <span aria-hidden="true">&times;</span>
              </button>
            )}
          </div>
          <div className="modal-body">
            {content}
          </div>
          <div className="modal-footer">
            {buttons.map((button, idx) =>
              // This is a pretty crude way to inject onClick handlers
              // in the buttons passed from the calling code; something
              // more flexible and sophisticated is left as exercise
              // for the application developer. :)
              React.cloneElement(button, {
                key: idx,
                onClick: () => { callback(button) },
              })
            )}
          </div>
        </div>
      </div>
    </div>
  );
};

export default Modal;