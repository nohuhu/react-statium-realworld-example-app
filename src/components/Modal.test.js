import React from 'react';
import { mount } from 'enzyme';

import Modal from './Modal.jsx';

describe("Modal component", () => {
  describe("rendering", () => {
    it("should render hidden modal and no backdrop", async () => {
      const tree = mount(
        <Modal show={false}
          title="foo bar"
          content="frobbe throbbe"
          buttons={[]}
        />
      );

      await sleep();

      tree.update();

      expect(tree).toMatchSnapshot();

      expect(document.body.classList.contains('modal-open')).toBe(false);

      const backdropEl = document.body.querySelector('.modal-backdrop');
      expect(backdropEl).toBe(null);
    });

    it("should render visible modal with backdrop", async () => {
      const tree = mount(
        <Modal show={true}
          title="krung klatz"
          content="froo froo funtz"
          buttons={[]}
        />
      );

      await sleep(10);

      tree.update();

      expect(tree).toMatchSnapshot();

      expect(document.body.classList.contains('modal-open')).toBe(true);

      const backdropEl = document.body.querySelector('.modal-backdrop');
      expect(backdropEl.className).toBe('modal-backdrop fade show');
    });
  });

  describe("buttons", () => {
    let callback;

    beforeEach(() => {
      callback = jest.fn();
    });

    it("should call the callback on close button click", () => {
      const tree = mount(
        <Modal show={true}
          title="plugh kludge"
          content="sorpo asto"
          buttons={[]}
          callback={callback}
        />
      );

      tree.find('button.close').simulate('click');

      expect(callback).toHaveBeenCalledWith(null);
    });

    it("should call the callback on footer button click", () => {
      const fooButton = <button className="foo">foo</button>;

      const tree = mount(
        <Modal show={true}
          title="jurg fropp"
          content="ordo wupp"
          buttons={[fooButton]}
          callback={callback}
        />
      );

      tree.find('button.foo').simulate('click');

      expect(callback).toHaveBeenCalledWith(fooButton);
    });
  });
});
