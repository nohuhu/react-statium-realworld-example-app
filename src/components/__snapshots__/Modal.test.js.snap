// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component rendering should render hidden modal and no backdrop 1`] = `
<Modal
  buttons={Array []}
  content="frobbe throbbe"
  show={false}
  title="foo bar"
>
  <div
    className="modal fade"
    role="dialog"
    tabIndex="-1"
  >
    <div
      className="modal-dialog"
      role="document"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <h5
            className="modal-title"
          >
            foo bar
          </h5>
          <button
            aria-label="Close"
            className="close"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </div>
        <div
          className="modal-body"
        >
          frobbe throbbe
        </div>
        <div
          className="modal-footer"
        />
      </div>
    </div>
  </div>
</Modal>
`;

exports[`Modal component rendering should render visible modal with backdrop 1`] = `
<Modal
  buttons={Array []}
  content="froo froo funtz"
  show={true}
  title="krung klatz"
>
  <div
    className="modal fade"
    role="dialog"
    tabIndex="-1"
  >
    <div
      className="modal-dialog"
      role="document"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <h5
            className="modal-title"
          >
            krung klatz
          </h5>
          <button
            aria-label="Close"
            className="close"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </div>
        <div
          className="modal-body"
        >
          froo froo funtz
        </div>
        <div
          className="modal-footer"
        />
      </div>
    </div>
  </div>
</Modal>
`;
